yt_dlp/postprocessor/mp4direct.py

```diff
@@ -0,0 +1,142 @@
+from math import inf
+
+from .common import PostProcessor
+from ..utils import prepend_extension
+
+from ..mp4_parser import (
+    write_mp4_boxes,
+    parse_mp4_boxes,
+    pack_be32,
+    pack_be64,
+    unpack_ver_flags,
+    unpack_be32,
+    unpack_be64,
+)
+
+
+class MP4FixupTimestampPP(PostProcessor):
+
+    @property
+    def available(self):
+        return True
+
+    def analyze_mp4(self, filepath):
+        """ returns (baseMediaDecodeTime offset, sample duration cutoff) """
+        smallest_bmdt, known_sdur = inf, set()
+        with open(filepath, 'rb') as r:
+            for btype, content in parse_mp4_boxes(r):
+                if btype == 'tfdt':
+                    version, _ = unpack_ver_flags(content[0:4])
+                    # baseMediaDecodeTime always comes to the first
+                    if version == 0:
+                        bmdt = unpack_be32(content[4:8])
+                    else:
+                        bmdt = unpack_be64(content[4:12])
+                    if bmdt == 0:
+                        continue
+                    smallest_bmdt = min(bmdt, smallest_bmdt)
+                elif btype == 'tfhd':
+                    version, flags = unpack_ver_flags(content[0:4])
+                    if not flags & 0x08:
+                        # this box does not contain "sample duration"
+                        continue
+                    # https://github.com/gpac/mp4box.js/blob/4e1bc23724d2603754971abc00c2bd5aede7be60/src/box.js#L203-L209
+                    # https://github.com/gpac/mp4box.js/blob/4e1bc23724d2603754971abc00c2bd5aede7be60/src/parsing/tfhd.js
+                    sdur_start = 8  # header + track id
+                    if flags & 0x01:
+                        sdur_start += 8
+                    if flags & 0x02:
+                        sdur_start += 4
+                    # the next 4 bytes are "sample duration"
+                    sample_dur = unpack_be32(content[sdur_start:sdur_start + 4])
+                    known_sdur.add(sample_dur)
+
+        maximum_sdur = max(known_sdur)
+        for multiplier in (0.7, 0.8, 0.9, 0.95):
+            sdur_cutoff = maximum_sdur * multiplier
+            if len(set(x for x in known_sdur if x > sdur_cutoff)) < 3:
+                break
+        else:
+            sdur_cutoff = inf
+
+        return smallest_bmdt, sdur_cutoff
+
+    def modify_mp4(self, src, dst, bmdt_offset, sdur_cutoff):
+        with open(src, 'rb') as r, open(dst, 'wb') as w:
+            def converter():
+                moov_over, in_secondary_moov = False, False
```

Comment:
this much nesting is confusing. Can't you move `convertor` out of the with block?